Core/PersistentPixel.swift

```diff
@@ -0,0 +1,326 @@
+            completion(PersistentPixelStorageError.readError(error))
+        }
+    }
+
+    // MARK: - Private
+
+    /// Sends queued pixels and calls the completion handler with any that failed.
+    private func fire(queuedPixels: [PersistentPixelMetadata], completion: @escaping ([PersistentPixelMetadata]) -> Void) {
+        let dispatchGroup = DispatchGroup()
+
+        let failedPixelsAccessQueue = DispatchQueue(label: "Failed Pixel Retry Attempt Metadata Queue")
+        var failedPixels: [PersistentPixelMetadata] = []
+        let currentDate = dateGenerator()
+
+        for pixelMetadata in queuedPixels {
+            if let originalSendDateString = pixelMetadata.timestamp,
+               let originalSendDate = dateFormatter.date(from: originalSendDateString),
+               let date28DaysAgo = calendar.date(byAdding: .day, value: -28, to: currentDate) {
+                if originalSendDate < date28DaysAgo {
+                    continue
+                }
+            } else {
+                // If we don't have a timestamp for some reason, ignore the retry - retries are only useful if they have a timestamp attached
+                continue
+            }
+
+            var pixelParameters = pixelMetadata.additionalParameters
+            pixelParameters[PixelParameters.retriedPixel] = "1"
+
+            dispatchGroup.enter()
+            pixelFiring.fire(
+                pixelNamed: pixelMetadata.pixelName,
+                forDeviceType: UIDevice.current.userInterfaceIdiom,
+                withAdditionalParameters: pixelParameters,
+                allowedQueryReservedCharacters: nil,
+                withHeaders: APIRequest.Headers(),
+                includedParameters: pixelMetadata.includedParameters,
+                onComplete: { error in
+                    if error != nil {
+                        failedPixelsAccessQueue.sync {
+                            failedPixels.append(pixelMetadata)
+                        }
+                    }
+
+                    dispatchGroup.leave()
+                }
+            )
+        }
+
+        dispatchGroup.notify(queue: .global()) {
+            completion(failedPixels)
+        }
+    }
+
+    private func save(_ pixelMetadata: PersistentPixelMetadata) throws {
+        pixelProcessingLock.lock()
+        defer { pixelProcessingLock.unlock() }
+
+        Logger.general.debug("Saving persistent pixel named \(pixelMetadata.pixelName)")
+
+        if self.isSendingQueuedPixels {
+            self.failedPixelsPendingStorage.append(pixelMetadata)
+        } else {
+            try self.persistentPixelStorage.append(pixel: pixelMetadata)
+        }
+    }
+
+    private func stopProcessingQueueAndPersistPendingPixels() {
+        pixelProcessingLock.lock()
+        defer { pixelProcessingLock.unlock() }
+
+        for pixel in self.failedPixelsPendingStorage {
+            try? self.persistentPixelStorage.append(pixel: pixel)
```

Comment:
Can we append as batch?

Reply:
This is resolved now, we no longer re-append failed pixels as the `failedPixelsPendingStorage` mechanism has been entirely removed.

We now collect a set of successfully-retried pixel UUIDs that get filtered out at the end of the retrying process. This means that incoming pixels can now be written to storage while processing is active, since we're no longer replacing the entire post-processing pixel array.